utils/vlog/vlog_to_pbtype.py

```diff
@@ -54,68 +54,225 @@
+        """
+        name = xml_tag.get("name")
+        numb = int(xml_tag.get("num_pins"))
+        if numb == 1:
+            return [name]
+        else:
+            return ["{}{}".format(name, i) for i in range(numb)]
+
+    # Load the SDF
+    with open(sdf_file_name, "r") as fp:
+        sdf = sdfparse.parse(fp.read())
+
+    # Get timescale
+    scale = parse_timescale(sdf["header"]["timescale"])
+
+    # Get pb name
+    pb_name = pb_type_xml.get("name")
+
+    # Get pb ports (inputs, clocks, outputs)
+    pb_inputs = []
+    for xml_tag in pb_type_xml.findall("input"):
+        pb_inputs.extend(pin_list(xml_tag))
+    pb_clocks = []
+    for xml_tag in pb_type_xml.findall("clock"):
+        pb_clocks.extend(pin_list(xml_tag))
+    pb_outputs = []
+    for xml_tag in pb_type_xml.findall("output"):
+        pb_outputs.extend(pin_list(xml_tag))
+
+    # No cell path given, try using the pb_name
+    if sdf_cell_paths is None:
+        sdf_cell_paths = (pb_name.upper())
+
+    # Collect timings for given paths
+    sdf_timings = collect_timings(sdf, sdf_cell_paths)
+    if len(sdf_timings) == 0:
+        exit(-1)
+
+    # Process SDF timing entires
+    for key, sdf_timing in sdf_timings.items():
+
+        sdf_inp = sdf_timing["from_pin"]
+        sdf_out = sdf_timing["to_pin"]
+
+        assert sdf_timing["is_incremental"] == False
+        assert sdf_timing["is_cond"] == False
+
+        # IOPATH delay
+        if sdf_timing["type"] == "iopath":
+
+            assert sdf_timing["is_absolute"] == True
+
+            # Find pins
+            pb_inp = sdf_inp if sdf_inp in pb_inputs else None
+            pb_clk = sdf_inp if sdf_inp in pb_clocks else None
+            pb_out = sdf_out if sdf_out in pb_outputs else None
+
+            # Cannot match SDF to <pb_type>
+            if (pb_inp is None and pb_clk is None) or pb_out is None:
+                print(
+                    "Cannot match pins ({} -> {}) for SDF timing entry:".
+                    format(sdf_inp, sdf_out)
+                )
+                print(" ", sdf_timing)
+                continue
+
+            # "delay_constant"
+            if pb_clk is None:
+                xml_tag = ET.SubElement(pb_type_xml, "delay_constant")
+                xml_tag.set("in_port", "{}.{}".format(pb_name, pb_inp))
+                xml_tag.set("out_port", "{}.{}".format(pb_name, pb_out))
+                for var in ("min", "max"):
+                    xml_tag.set(
+                        var, "{:.3e}".format(
+                            sdf_timing["delay_paths"][sdf_variant][var] * scale
+                        )
+                    )
+
+            # "T_clock_to_Q"
+            else:
+                xml_tag = ET.SubElement(pb_type_xml, "T_clock_to_Q")
+                xml_tag.set("clock", "{}.{}".format(pb_name, pb_clk))
+                xml_tag.set("port", "{}.{}".format(pb_name, pb_out))
+                for var in ("min", "max"):
```

Comment:
this loop is identical for both the cases. Please make it common for both of them

Reply:
Done.